### ISSUE TITLE
Extend transaction subscription to include block_time

### ISSUE THREAD

**Issue by vovkman** [rpcpool/yellowstone-grpc#228]
Currently the transaction notification doesn't include `block_time`, however this is included in the standard RPC interface [here](https://docs.solana.com/api/http#gettransaction). A few options

1. Update `SubscribeUpdateTransaction` to include `optional uint64 block_time` field that gets set based on existence of block_meta in the `SlotsMessages`
2.  Same as 1 but update `SubscribeRequestFilterTransactions` with `optional bool block_meta_required` field that defaults to false

if `block_meta_required` is true
  - Prevent txn broadcast until `block_meta` is present
 

else
  - broadcast whether `block_meta` is present or not

**Comment by vovkman** [comment]
cc @fanatid

**Comment by kespinola** [comment]
Giving this a look today. I'll report back on what I find.

**Comment by linuskendall** [comment]
Wouldnt it make more sense to just subscribe to the block meta updates
instead of having some side effect that if you provide this param that the
TX transmission gets delayed?

That seems like an "unexpected" side effect that could trip someone up ?

I understand that the goal is to avoid to get multiple different types of
messages for the user to handle?

**Comment by kespinola** [comment]
@vovkman the block time is made available in the slot update event but not the transaction. It does have the slot in which case getBlockTime can be passed for fined the block_time.

https://docs.solana.com/api/http#getblocktime

**Comment by linuskendall** [comment]
No need for getBlockTime, the block subscription data already provides this via gRPC :
https://github.com/rpcpool/yellowstone-grpc/blob/8f1cc8bdf411b6f495213dfcf371a5c8e23b2f70/yellowstone-grpc-proto/proto/solana-storage.proto#L7

**Comment by linuskendall** [comment]
This subscription would give access to both transactions and block_time since the block meta message would include the block time for the specific transactions:

```
{"slots": {}, "accounts": {}, "transactions":  { "serum": { "vote": false, "failed": false, "account_include": [ "9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin" ]}}, "blocks": {}, "blocks_meta": { "blockmetadata": {} }, "accounts_data_slice": []}
```

**Comment by vovkman** [comment]
Yeah this is meant to simplify things for clients. I think the `wait_for_block_meta` flag is a good idea, and it can be clearly called out in docs what it does

**Comment by kespinola** [comment]
@vovkman would a client cache that buffered and then flushed the transactions once the block be confirmed suffice for your  use case? This would allow you to have access to the block metadata at the same time as processing the transaction but does add delay to the processing.

**Comment by vovkman** [comment]
I would prefer for the changes to be in the plugin itself

**Comment by vovkman** [comment]
I am personally more in favor of `Wait for block flag`, the pros of the second option can still be leveraged if clients want it as a separate feature. The cons about persistence/crashing is present for all subscriptions currently, and perf can be tested, I don't suspect any issue since the plugin buffers the last 10 SlotMessages already anyways

**Comment by linuskendall** [comment]
Just to follow up on this, would the following make sense:

Instead of "wait for block flag" add something like an additional commitment level (processed_complete?) which would have the same semantics as websockets processed, ie. it would only return one record per slot (max(write_version)) and would return the record with block meta information as block time.

In this way the plugin only buffers a single write update and flushes the last `write_version` along with the block time at the time when block_meta is received? 

This mode would mimic the websockets and would also

**Comment by linuskendall** [comment]
> I am personally more in favor of `Wait for block flag`, the pros of the second option can still be leveraged if clients want it as a separate feature. The cons about persistence/crashing is present for all subscriptions currently, and perf can be tested, I don't suspect any issue since the plugin buffers the last 10 SlotMessages already anyways

We will need to buffer all the transactions for the slot and flush them so it definitely means a lot more server side buffering for the client in this mode (however: confirmed and finalzied anyway has the same semantics so I don't think it will be a huge difference).

I think having a flag is a little bit inelegant, but maybe it's the best way to do it? another way would be to specify a unique commitment level (`processed_complete`?) which would do two things:

- for transactions, buffer the transactions until the blockmeta is available and flush them with the blockmeta included
- for accounts, buffer the latest account writes (by write version) and flush one account write per block with the blockmeta incldued
